### DISCUSSION TITLE
Embed avalonia in a native android/ios app

### DISCUSSION THREAD

**Discussion by ILAgent** [AvaloniaUI/Avalonia#10305] [Q&A]
Is it possible to embed Avalonia in an existing native (kotlin/swift) app? 
And vice versa, how to use native libraries in an Avalonia project?

**Comment by emmauss** [comment]
You can host any Avalonia ui in an `AvaloniaView`, which is a native platform view. The Avalonia system itself must be initialized first with the preferred themes, but after that, you can add `AvaloniaView` to any viewgroup.

You can use native Xamarin.Android/.Net Android libraries in Avalonia as long as the project's target framework supports it. Though if you want to host native Android views in Avalonia on Android, you'd have to use `AndroidViewControlHandle`.

**Reply by ILAgent** [reply]
Thank you. 
I mean truly native apps) For instance I have an android app written in Kotlin. And I'd like to add there a screen written with Avalonia. And after that reuse the screen in a native (swift) Ios  app. Is it possible? I know about a stuff Embeddinator 4000 but that's unusable.

**Reply by maxkatz6** [reply]
That's more an area of .NET runtime questions, as we don't maintain interop between languages. 

In general, it should be possible to use Kotlin and Swift from .NET
https://learn.microsoft.com/en-us/xamarin/android/platform/binding-kotlin-library/
https://learn.microsoft.com/en-us/xamarin/ios/platform/binding-swift/walkthrough

Also probably more specific
https://learn.microsoft.com/en-us/xamarin/ios/platform/binding-swift/
https://learn.microsoft.com/en-us/xamarin/android/platform/binding-java-library/

Not sure how well it works though with .NET 6.

**Reply by maxkatz6** [reply]
No idea about the opposite way though, using .NET runtime from swift/kotlin apps.

**Reply by emmauss** [reply]
> That's more an area of .NET runtime questions, as we don't maintain interop between languages.
> 
> In general, it should be possible to use Kotlin and Swift from .NET https://learn.microsoft.com/en-us/xamarin/android/platform/binding-kotlin-library/ https://learn.microsoft.com/en-us/xamarin/ios/platform/binding-swift/walkthrough
> 
> Not sure how well it works though with .NET 6.

Binding works fine on Net 6.0+. Compatibility depends on the library being bound, but with some effort, you can import most android native libraries to .Net Android.

**Reply by ILAgent** [reply]
> In general, it should be possible to use Kotlin and Swift from .NET

Thank you, just as I thought Avalonia relies on Xamarin in that case. 

> No idea about the opposite way though, using .NET runtime from swift/kotlin apps.

Ok, I was hoping something new was introduced in that area)

**Comment by tkefauver** [comment]
@emmauss is there any way to initialize Avalonia after startup? Like in Xamarin.Forms you can just call [`Init()`](https://devblogs.microsoft.com/xamarin/ios-extension-xamarin-forms/#how-does-it-work) at some point but I don't see how to do that w/ Avalonia. I'm trying to show Avalonia views from an ios keyboard extension.

EDIT: I freaking got it finally! The key was giving the extension project its own `AppDelegate` and `App` classes then manually calling `FinishedLaunching`. Also had to wrap all `AvaloniaView` calls in try/catch blocks for unknown reasons...
